ARCHITECTURE.md

```diff
@@ -0,0 +1,275 @@
+  path: 
+   [ 1, 2, 0, 1, 2, 0, 2, 2, 3, 0, 1, 2, 1, 3, 0, 3, 0, 0, 2, 1, 0, 2, 0, 0, 2, 0, 0, 3, 2, 1, 1, 2,
+     0, 1, 2, 3, 2, 2, 2, 0, 3, 1, 1, 3, 0, 3, 1, 3, 0, 1, 0, 1, 3, 2, 0, 2, 2, 3, 2, 2, 3, 3, 2, 3,
+     4 ] } 
+```
+
+If you compare this path to the one for `/a/b/c` above, you'll see that the
+first 64 2-bit characters match. This is because `/a/b` is a prefix of `/a/b/c`.
+
+Since this is the first entry, `seq` is 0. Since this is the only known feed,
+`feed` is also 0. `feeds` is an array of entries of the form `{ key: Buffer,
+seq: Number }` that let you map the numeric value `feed` to a hypercore key and
+its sequence number head. `feeds` isn't always set: it only gets included when
+it changes compared to `node.seq - 1`, in the interest of storing less data per
+node.
+
+Now we write `/a/c = hello` and get this node:
+
+```js
+{ key: '/a/c',
+  value: 'hello',
+  clock: [ 0 ],
+  trie: [ , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , [ , , [ { feed: 0, seq: 0 } ] ] ],
+  feeds: [],
+  feedSeq: 0,
+  feed: 0,
+  seq: 1,
+  path: 
+   [ 1, 2, 0, 1, 2, 0, 2, 2, 3, 0, 1, 2, 1, 3, 0, 3, 0, 0, 2, 1, 0, 2, 0, 0, 2, 0, 0, 3, 2, 1, 1, 2,
+     0, 1, 1, 0, 1, 2, 3, 2, 2, 2, 0, 0, 3, 1, 2, 1, 3, 3, 3, 3, 3, 3, 0, 3, 3, 2, 3, 2, 3, 0, 1, 0,
+     4 ] }
+```
+
+As expected, this node has the same `feed` value as before (since we're only
+writing to one feed). Its `seq` is 1, since the last was 0. Notice that `feeds`
+isn't included, because the mapping of the numeric `feed` value to a key hasn't
+changed.
+
+Also, this and the previous node have the first 32 characters of their `path` in
+common (the prefix `/a`).
+
+Notice though that `trie` is set. It's a long but sparse array. It has 35
+entries, with the last one referencing the first node inserted (`a/b/`). Why?
+
+(Really, it's a sparse array with 64 entries (the length of the `path`). But in
+Javascript only the entries preceding the last value get printed.)
+
+If you visually compare this node's `path` with the previous node's `path`, how
+many entries do they have in common before the 2-bit numbers diverge?
+
+At the 35th entry.
+
+What this is saying is "if the hash of the key you're looking for differs from
+mine on the 35th entry, you want to travel to `{ feed: 0, seq: 0 }` to find the
+node you're looking for.
+
+This is how finding a node works, starting at any other node:
+
+1. Compute the 2-bit hash sequence of the key you're after (e.g. `a/b`)
+2. Lookup the newest entry in the feed.
+3. Compare its `path` against the hash you just computed.
+4. If you discover that the `path` and your hash match, then this is the node
+   you're looking for!
+5. Otherwise, once a the 2-bit character from `path` and your hash disagree,
+   node the index # where they differ and look up that value in the node's
+   `trie`. Fetch that node at the given feed and sequence number, and go back to
+   step 2. Repeat until you reach step 3 (match) or there is no entry in the
```

Comment:
go back to step 3? Repeat until you reach step 4?

Reply:
You're right! At one point I removed some steps to make it simpler and forgot to update the references.